File: solutions/0084-largest-rectangle-in-histogram/solution.cpp

class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        //store next smaller elements for left to right - 

        vector<int> leftNextSmaller(heights.size(),-1);
        vector<int> rightNextSmaller(heights.size(),heights.size());

        stack<int> stk;
        for(int i=0;i<heights.size();i++){
            while(!stk.empty() and heights[i]<=heights[stk.top()]){
                stk.pop();
            }
            if(!stk.empty()){
                leftNextSmaller[i] = stk.top();
            }
            stk.push(i);
        }
        while(!stk.empty()){
            stk.pop();
        }
        for(int i=heights.size()-1;i>=0;i--){
            while(!stk.empty() and heights[i]<=heights[stk.top()]){
                stk.pop();
            }
            if(!stk.empty()){
                rightNextSmaller[i] = stk.top();
            }
            stk.push(i);
        }  
        int maxarea = 0;
        for(int i=0;i<heights.size();i++){
            int area =(rightNextSmaller[i]-leftNextSmaller[i]-1)*heights[i];
            maxarea = max(maxarea,area);
        }
        return maxarea;
    }
};

File: solutions/0084-largest-rectangle-in-histogram/solution.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        # precompute next smaller 
        nextSmallerIndex = [len(heights)]*len(heights)
        prevSmallerIndex = [-1]*len(heights)

        stk = []
        for i in range(0,len(heights)):
            if not stk or heights[stk[-1]]<=heights[i]:
                stk.append(i)
                continue
            # pop until top is lesser
            while stk and heights[stk[-1]]>heights[i]:
                idx=stk.pop()
                # current i is the next smaller for poped indexes
                nextSmallerIndex[idx]=i
            stk.append(i)

        # precompute prevsmaller idex
        stk = []
        for i in range(len(heights)-1,-1,-1):
            if not stk or heights[stk[-1]]<=heights[i]:
                stk.append(i)
                continue
            while stk and heights[stk[-1]]>heights[i]:
                idx = stk.pop()
                prevSmallerIndex[idx]=i
            stk.append(i)
        
        # compute largest rectange

        maxarea = 0
        for i in range(0,len(heights)):
            area = heights[i]*(nextSmallerIndex[i]-prevSmallerIndex[i]-1)
            maxarea = max(maxarea,area)
        return maxarea


        
